src/bun.js/bindings/exports.zig

```diff
@@ -1031,6 +1048,343 @@ pub const ZigConsoleClient = struct {
+                    .single_line = true,
+                    .max_depth = 5,
+                },
+            };
+        }
+
+        /// Compute how much horizontal space will take a JSValue when printed
+        fn getWidthForValue(this: *TablePrinter, value: JSValue) u32 {
+            var counting_writer = std.io.countingWriter(std.io.null_writer);
+            var value_formatter = (&this.value_formatter).*;
+            value_formatter.format(
+                ZigConsoleClient.Formatter.Tag.get(value, this.globalObject),
+                @TypeOf(counting_writer).Writer,
+                counting_writer.writer(),
+                value,
+                this.globalObject,
+                false,
+            );
+            return @as(u32, @intCast(counting_writer.bytes_written));
+        }
+
+        /// Update the sizes of the columns for the values of a given row, and create any additional columns as needed
+        fn updateColumnsForRow(this: *TablePrinter, columns: *std.ArrayList(Column), row_key: RowKey, row_value: JSValue) !void {
+            // update size of "(index)" column
+            const row_key_len: u32 = switch (row_key) {
+                .str => |value| @intCast(value.length()),
+                .num => |value| @intCast(std.fmt.count("{d}", .{value})),
+            };
+            columns.items[0].width = @max(columns.items[0].width, row_key_len);
+
+            // special handling for Map: column with idx=1 is "Keys"
+            if (this.jstype.isMap()) {
+                const entry_key = row_value.getIndex(this.globalObject, 0);
+                const entry_value = row_value.getIndex(this.globalObject, 1);
+                columns.items[1].width = @max(columns.items[1].width, this.getWidthForValue(entry_key));
+                this.values_col_width = @max(this.values_col_width orelse 0, this.getWidthForValue(entry_value));
+                return;
+            }
+
+            if (row_value.isObject()) {
+                // object ->
+                //  - if "properties" arg was provided: iterate the already-created columns (except for the 0-th which is the index)
+                //  - otherwise: iterate the object properties, and create the columns on-demand
+                if (!this.properties.isUndefined()) {
+                    for (1..columns.items.len) |i| {
+                        if (row_value.getWithString(this.globalObject, columns.items[i].name)) |value| {
+                            columns.items[i].width = @max(columns.items[i].width, this.getWidthForValue(value));
+                        }
+                    }
+                } else {
+                    var cols_iter = JSC.JSPropertyIterator(.{
+                        .skip_empty_name = false,
+                        .include_value = true,
+                    }).init(this.globalObject, row_value.asObjectRef());
+                    defer cols_iter.deinit();
+
+                    while (cols_iter.next()) |col_key| {
```

Comment:
sidenote: in a different PR, we should change PropertyIterator to use `bun.String` because the memory here is pretty unsafe (its referencing the latin1/utf16 character pointer which may be freed at any time later)

Reply:
this is a problem elsewhere, not here. and in this case, it's pretty much fine because the values will be alive for the scope of this function call anyway